plugins/algoltek-usb/fu-algoltek-usb-device.c

```diff
@@ -0,0 +1,633 @@
+
+		fu_byte_array_set_size(partFirmware, transferDataSize, 0);
+
+		for (guint i = 0; i < transferDataSize; i++)
+			partFirmware->data[i] = firmwareData->data[i + startIndex];
+
+		if (!g_usb_device_control_transfer(usb_device,
+						   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+						   G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+						   G_USB_DEVICE_RECIPIENT_INTERFACE,
+						   ALGOLTEK_USB_REQUEST_CMD09,
+						   value, /* value */
+						   index, /* idx */
+						   partFirmware->data,
+						   partFirmware->len,
+						   NULL,
+						   ALGOLTEK_DEVICE_USB_TIMEOUT,
+						   NULL,
+						   error)) {
+			g_prefix_error(error, "failed to contact device: ");
+			return FALSE;
+		}
+
+		startIndex += transferDataSize;
+		restTransferDataSize -= transferDataSize;
+		if (countCheck == (256 / maxPktSize) || restTransferDataSize == 0) {
+			countCheck = 0;
+			result = fu_algoltek_device_status_check(self, error);
+			if (!result)
+				return FALSE;
+		}
+		fu_progress_set_percentage_full(progress, startIndex, firmwareData->len);
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_usb_device_setup(FuDevice *device, GError **error)
+{
+	FuAlgoltekUsbDevice *self = FU_ALGOLTEK_USB_DEVICE(device);
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(device));
+	g_autoptr(GByteArray) versionData = NULL;
+	g_autofree gchar *version_str = NULL;
+
+	/* UsbDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_algoltek_usb_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	versionData = fu_algoltek_device_CMD01(self, error);
+	version_str = g_strdup_printf("%s", versionData->data);
+
+	g_assert(self != NULL);
+	g_assert(usb_device != NULL);
+	fu_device_set_version(device, version_str);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_usb_device_write_firmware(FuDevice *device,
+				      FuFirmware *firmware,
+				      FuProgress *progress,
+				      FwupdInstallFlags flags,
+				      GError **error)
+{
+	FuAlgoltekUsbDevice *self = FU_ALGOLTEK_USB_DEVICE(device);
+	g_autoptr(GBytes) blob_ISP = NULL;
+	g_autoptr(GBytes) blob_payload = NULL;
+	g_autoptr(GByteArray) fw_ISP = g_byte_array_new();
+	g_autoptr(GByteArray) fw_payload = g_byte_array_new();
+	gboolean result;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 15, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 95, NULL);
+
+	result = fu_algoltek_device_CMD03(self, error);
+	if (!result) {
+		g_prefix_error(error, "Error Code 0x09 : ");
+		return FALSE;
+	}
+
+	result = fu_algoltek_device_CMD04(self, 2, error);
+	if (!result) {
+		g_prefix_error(error, "Error Code 0x20_2 : ");
+		return FALSE;
+	}
+
+	fu_device_sleep(FU_DEVICE(self), 900);
+
+	result = fu_algoltek_device_CMD05(self, 0x80AD, 0, error);
+	if (!result) {
+		g_prefix_error(error, "Error Code 0x07_80AD : ");
```

Comment:
no space before the `:` and also the error message make no sense at all

Reply:
Text content updated.